rfcs/0040-rosetta-timelocking.md

```diff
@@ -0,0 +1,115 @@
+
+## Motivation
+
+[motivation]: #motivation
+
+Right now, Rosetta can not handle accounts with time locks on them. The specification demands we present the liquid balance in the accounts. Unfortunately, in the protocol we sample the liquidity curve at the moment funds are attempting to be transferred which is at odds with how Rosetta attempts to understand the world. We wish to support this.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+### Balance exemptions and historical balance lookups
+
+Rosetta supports the notion of a balance exempt account. These are accounts that one should not use operations alone to monitor changes in balance. The specification details that this should be used sparingly, but goes on to suggest that vesting accounts are one such example. If we only support time-locked accounts in the genesis ledger, all we need to do is at Rosetta-server-startup-time is pull those time-locked account addresses and fill the balance exemption field. If we don't, we'll need to update this field dynamically, it remains to be seen if this is allowable by the Rosetta specification, see unresolved question 1.
+
+This is not a solution alone -- the specification goes on to say that in the case that one or more accounts are balance exempt, you must support historical balance lookups. This is difficult for us because Mina's constant-sized blockchain does not contain historical data. We use the archive node to store extra data for us -- we're already using it to store all the blocks that our nodes see. In order to support historical balance lookups, we have to add extra information so we can compute this data. Specifically, we'll add information about the current balance of any accounts touched during a transaction whenever we store in the archive node. We can use this and the genesis ledger from Rosetta to reconstruct the current balance of any account at any block by crawling backward from that block until we see a transaction (or fallback to the genesis ledger if no such transaction exists). Additionally, we can use the extra timing information in time-locked accounts in the genesis ledger to sample the liquidity curve at any moment as well. In the universe where we need to support time-locked account creation after genesis, we should be able to extract the timing information from the relevant transaction that creates the account via the archive node as well.
+
+### Implementation details
+
+In some cases, two separate approaches are outlined. This RFC proposes that we go with Approach A in the case that by mainnet we do not need to support creating time-locked accounts after the genesis ledger, and Approach B if we do end up needing to support this feature.
+
+**Protocol/Archive**
+
+- Add the following columns to the `blocks_user_commands` table in SQL
+
+```
+fee_payer_amount : bigint NOT NULL,
+sender_amount : bigint NOT NULL,
+receiver_amount :  bigint NOT NULL
```

Comment:
I think these should be `fee_payer_balance`/`sender_balance`/`receiver_balance`, to match our parlance elsewhere.